### PR TITLE
Call `get_pending_validators`

### LINKED ISSUE #49: Feature request for AuRa: using `getPendingValidators()` getter instead of `InitiateChange` event
We agreed that we'll use the ~~`getValidators()`~~ [`getPendingValidators()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/abstracts/ValidatorSetBase.sol#L192) getter instead of catching the `InitiateChange` event. Parity should just read the current validator set with ~~`getValidators()`~~ `getPendingValidators()` every block, and if it discovers that the set is changed in the contract, it launches changing of the set in the engine.

The [`finalizeChange()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/abstracts/ValidatorSetBase.sol#L85) function (described in https://wiki.parity.io/Validator-Set.html) should be called by Parity after [`newValidatorSet()`](https://github.com/poanetwork/pos-contracts/blob/f7d080fbf5e16bd9b8cdb474bf9541bfe292c796/contracts/ValidatorSetAuRa.sol#L55) is called and the internal list of validators is updated in engine (this is default behavior for AuRa in Parity, so I think the code of the engine shouldn't be changed for that).

**UPDATE:** stroke out things below are already implemented by the contracts.

~~Parity must call [`newValidatorSet()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/ValidatorSetAuRa.sol#L41) function by `SYSTEM_ADDRESS` once a week.~~

~~Since in AuRa the block time is determined (5 seconds), the `ValidatorSet` contract will keep track of elapsed time by itself, so there is no need to measure 1 week in Parity.~~

~~1 week = 120960 blocks~~

~~Parity must call the [`newValidatorSetCallable()`](https://github.com/poanetwork/pos-contracts/blob/1240a2e6ab098f18c868324fb2c22454417807d5/contracts/ValidatorSetAuRa.sol#L96) getter every block and check whether to call `newValidatorSet`.~~

~~So, if `newValidatorSetCallable()` returns `true`, the `newValidatorSet()` must be called at the end of the current block.~~

~~**Important note:** transaction of calling `newValidatorSet()` must be included in a block before transaction of calling `BlockReward.reward()`.~~

### PR DESCRIPTION
**Pull request by DemiMarie**
instead of relying on the `InitiateChange` event.

Closes #49.

### CODE REVIEW

**Comment by DemiMarie** [comment]
@afck this is not backwards-compatible with existing contracts.

**Comment by vkomenda** [comment]
**Note:** This PR is supposed to fix #49. Refer there for a description of changes.

File: ethcore/src/engines/authority_round/mod.rs [hunk 1]
```diff
@@ -83,8 +83,6 @@ pub struct AuthorityRoundParams {
 	pub maximum_empty_steps: usize,
 	/// Transition block to strict empty steps validation.
 	pub strict_empty_steps_transition: u64,
-	/// Sets whether Aura will use Proof of Authority (PoA) or Proof of Stake (PoS) consensus.
-	pub consensus_kind: ConsensusKind,
```
Comment (vkomenda) [inline ethcore/src/engines/authority_round/mod.rs]: Why is this removed?

**Comment by afck** [comment]
> this is not backwards-compatible with existing contracts.

Exactly: Shouldn't we create a new kind of `validator_set` so that the old one still works in existing networks?

**Comment by vkomenda** [comment]
[Kovan](https://github.com/parity-contracts/kovan-validator-set/blob/8a7906a7ec8a3ccb8a02fb59576fa6e407386eaf/contracts/interfaces/ValidatorSet.sol#L29) requires `InitiateChange`, for example. Can `get_pending_validators` be additional functionality rather than a replacement?

File: ethcore/src/engines/authority_round/mod.rs (hunk 1 above)
Comment (DemiMarie) [reply ethcore/src/engines/authority_round/mod.rs]: It was dead and caused compiler warnings.

**Comment by c0gent** [comment]
@DemiMarie, the changes here and the changes in https://github.com/poanetwork/parity-ethereum/commit/1323ef7668bac6ba2be0e772d254f721de2fb138 do not account for the fact that Aura may be used in either PoA or PoS mode.  The `consensus_kind` parameter is what specifies this. The PoA functionality should be an addition, not a replacement.

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 2]
```diff
@@ -216,61 +215,14 @@ impl ValidatorSafeContract {
...
-		// only last log is taken into account
-		match decoded_events.next() {
-			None => None,
-			Some(matched_event) => Some(SimpleList::new(matched_event.new_set))
-		}
+	fn extract_from_event(&self, caller: Box<Call>) -> Option<SimpleList> {
+		let data = validator_set::functions::get_pending_validators::encode_input();
+		caller(self.contract_address, data)
+			.ok()
+			.and_then(|x|validator_set::functions::get_pending_validators::decode_output(&x.0).ok())
```
Comment (afck) [inline ethcore/src/engines/validator_set/safe_contract.rs]: Let's put a space after `|x|` (the old code had that, too).

File: ethcore/src/engines/validator_set/safe_contract.rs [hunk 3]
```diff
@@ -216,61 +215,14 @@ impl ValidatorSafeContract {
...
-			.filter_map(|log| {
-				validator_set::events::initiate_change::parse_log((log.topics.clone(), log.data.clone()).into()).ok()
-			});
-
-		// only last log is taken into account
-		match decoded_events.next() {
-			None => None,
-			Some(matched_event) => Some(SimpleList::new(matched_event.new_set))
-		}
+	fn extract_from_event(&self, caller: Box<Call>) -> Option<SimpleList> {
```
Comment (afck) [inline ethcore/src/engines/validator_set/safe_contract.rs]: The method docs are out of date now. Also, maybe it should be called `extract_from_block`, since we're not using an event anymore.

**Comment by DemiMarie** [comment]
> @DemiMarie, the changes here and the changes in [1323ef7](https://github.com/poanetwork/parity-ethereum/commit/1323ef7668bac6ba2be0e772d254f721de2fb138) do not account for the fact that Aura may be used in either PoA or PoS mode. The `consensus_kind` parameter is what specifies this. The PoA functionality should be an addition, not a replacement.

Agreed.  This should probably be a new type of `ValidatorSet`.  I can refactor the common components out later.

**Comment by DemiMarie** [comment]
Is it okay to add additional `ValidatorSet` kinds in [json/src/spec/validator_set.rs](https://github.com/poanetwork/parity-ethereum/blob/789bb9c852429594eff3aa194b6333eca01bd30b/json/src/spec/validator_set.rs)?

**Comment by afck** [comment]
Yes, I imagine that's the way to go. That way we keep supporting existing kinds of validator sets.
Regarding the names I'm not particularly creative either. :smile: `PosContract`? `CommitteeContract`?

**Comment by DemiMarie** [comment]
@afck @vkomenda is this good to merge?
